add support for timeout feature

kristopherbullinger opened the pull request:
Hello. Discord released a new feature recently allowing moderators/admins to time out users for a specified amount of time. Timing out a user requires a [new permission](https://discord.com/developers/docs/topics/permissions#permissions) and is part of the [Modify Guild Member](https://discord.com/developers/docs/resources/guild#modify-guild-member) API.

Thank you for your time reading and reviewing this pull request. I am happy to address any issues if you find that there are any.

nickelc commented on src/client/mod.rs:
```diff
@@ -369,7 +369,6 @@ impl<'a> Future for ClientBuilder<'a> {
 
             self.fut = Some(Box::pin(async move {
                 let url = Arc::new(Mutex::new(http.get_gateway().await?.url));
-
```
This change is unrelated.

kristopherbullinger replied:
I will undo this change. My mistake, left over from some debugging during local testing.

GnomedDev commented on src/builder/edit_member.rs:
```diff
@@ -87,4 +90,40 @@ impl EditMember {
+    /// Requires the [Moderate Members] permission.
+    ///
+    /// [Moderate Members]: crate::model::permission::Permissions::MODERATE_MEMBERS
+    pub fn remove_timeout(&mut self) -> &mut Self {
```
Seems like a weird name, maybe the other methods should take Option to follow the API or this should be renamed to `enable_communication`

kristopherbullinger replied:
I will update this method name.

GnomedDev reviewed (commented):
This PR also doesn't seem to add the new `communication_disabled_until` field to Member, however it may be missing something.

kristopherbullinger committed:
Add support for timeout feature

kristopherbullinger committed:
Update EditMember timeout methods to contain verbs

kristopherbullinger committed:
Add `communication_disabled_until` fields to Member and PartialMember

kristopherbullinger committed:
Add timeout method to Member

kristopherbullinger committed:
Add `communication_disabled_until` field to Member, PartialMember in tests

kristopherbullinger committed:
Remove unnecessary type parameter from `disable_communication_until`

kristopherbullinger commented:
In total, I have:
- Updated `Member` and `PartialMember` to contain a new `communication_disabled_until: Option<DateTime<Utc>>` field.
- Added `disable_communication_until_datetime` and `enable_communication` methods to the `Member` struct.
- Added three new methods to `EditMember`: `disable_communication_until`, `disable_communication_until_datetime`, and `enable_communication`.

I am wondering if it is better for the `EditMember::disable_communication_until_datetime` to accept as a parameter `DateTime<Utc>` rather than `DateTime<Tz> where Tz: TimeZone, Tz::Offset: Display` to simplify the API.

kristopherbullinger committed:
Update `enable_communication` to set `communication_disable_until` field to Null

kristopherbullinger committed:
Move `Member::disable_communication_until_datetime` into alphabetical position among methods

kristopherbullinger committed:
Add `Member::enable_communcation` method for removing timeout

kristopherbullinger committed:
Update timeout methods' documentation to reflect restrictions on timeout duration

kristopherbullinger commented on src/model/guild/member.rs:
```diff
@@ -92,7 +97,7 @@ impl Member {
 
                 Ok(())
-            },
+            }
```
These changes kept getting applied every time I ran RustFmt. Let me know if you want me to undo them.

arqunis replied:
These commas are meant to be there, as [defined by our rustfmt configuration](https://github.com/serenity-rs/serenity/blob/current/rustfmt.toml#L11). Which version of rustfmt are you using? We only support nightly, because we use options not yet supported in stable rustfmt.

kristopherbullinger replied:
I see. I was not using nightly -- I was letting an editor plugin do the formatting for me, which was using the default rustfmt channel. Running `cargo +nightly fmt` resolved the commas and other formatting discrepancies. Thank you.

kristopherbullinger committed:
Run cargo +nightly fmt

kristopherbullinger committed:
Fix unresolved doc link for permissions

arqunis commented:
Thanks!
